### DISCUSSION TITLE
Eloquent changes to serialize and casts

### DISCUSSION THREAD

**Discussion by DarkGhostHunter** [laravel/framework#39850] [Ideas]
Currently the Eloquent Model it's getting very cumbersome ~~to maintain~~ to solve the problem of mutability, given the approach of _mutators_ and getters.

My proposition is to separate attribute casting and attribute serializing, making it a two-way layer, allowing for more granular control of what interacts with the database.

```

             ┌────────────────────────────────────────────────────────────┐
             │                           Model                            │
┌──────────┐ │ ┌───────────────┐      ┌────────────┐      ┌─────────────┐ │
│          │ │ │  Serializer   │      │            │      │ Caster      │ │
│          │ │ │               │      │            │      │             │ │
│ Database─┼─┼─► unserialize() ├──────► Attributes ├──────┤ get()       │ │
│          │ │ │               │      │            │      │             │ │
│          ◄─┼─┤  serialize()  ◄──────┤            ◄──────┤ set()       │ │
└──────────┘ │ └───────────────┘      └────────────┘      │             │ │
             │                                            └─────────────┘ │
             └────────────────────────────────────────────────────────────┘

```

* A **serializer** is a class that is in direct contact with the database. It changes how the data is presented after is retrieved, and how it's serialized for saving. The serialize always returns primitives `bool`, `int`, `string`. If it's an `array`, then it should replace or update the array of attributes being serialized.

```php
use Illuminate\Contracts\Database\Eloquent\Serializer;
use Illuminate\Contracts\Database\Eloquent\Unserializer;

class Encrypt implements Serializer, Unserializer
{
    public function serialize(Model $model, string $key, mixed $value, array $attributes): int|string|bool|array
    {
        // ...
    }

    public function unserialize(Model $model, string $key, mixed $value, array $attributes): int|string|bool|array
    {
        // ...
    }
}
```

* A **caster** is a class that changes how the data is retrieved from the attributes array, pretty much like now mutators and getters works in Laravel. The `Mutable` contract will automatically "cache" any object returned to make it mutable, so `$model->collection->push('foo')` can work.

```php
use Illuminate\Contracts\Database\Eloquent\Setter;
use Illuminate\Contracts\Database\Eloquent\Getter;
use Illuminate\Contracts\Database\Eloquent\Mutable;

class Encrypt implements Setter, Getter, Mutable
{
    public function get(Model $model, string $key, mixed $value, array $attributes): mixed
    {
        // ...
    }

    public function set(Model $model, string $key, mixed $value, array $attributes): mixed
    {
        // ...
    }
}
```

This makes the `attributes` array become a representation of the truth, and the serializer the responsible of altering the truth so the database receive correct values. The serializer runs only when retrieving/saving.

It should be noted that a model should have a `$serializers` property to serialize attributes to the database. For example, decryption/encryption would be a serializer, while a Carbon object would be a Caster. You could mix both to encrypt a datetime.

```php
protected array $casts = [
    'greocery_list' => 'array:object',
    'submitted_at' => 'datetime',
];

protected array $serializers = [
    'submitted_at' => 'encrypt'
];
```

In any case, this would break a lot of code in Laravel, so this kind of implementation should be tentative for Laravel 10.

**Comment by spetz83** [comment]
It's not clear to me what problem you are attempting to solve with this. You say models are getting very cumbersome to maintain, but you do not provide any example to back up that claim. Is this a "you" problem, or a problem a bunch of users are having?

**Reply by DarkGhostHunter** [reply]
I updated it. It's not that' is cumbersome, but the implementation could be cleaner in my opinion. Currently, all getters and setters works on-demand, meaning, these cast and serialize at the same time.

I think that using the serialize/cast approach allows for greater flexibility when dealing with data being persisted from/to the database.

**Reply by DarkGhostHunter** [reply]
I'll add this also would get rid of the "cast cache".
